components/eamxx/src/physics/mam/eamxx_mam_aci_process_interface.cpp

```diff
@@ -152,75 +207,61 @@ void MAMAci::run_impl(const double dt) {
       rho_(icol,kk) = p_mid_(icol,kk)/(rair*T_mid_(icol,kk));
       w0_(icol,kk) = -1.0*omega_(icol,kk)/(rho_(icol,kk)*gravit);
-    }
-  }
-
-    /*
-    NOTE: All the inputs are available to compute w0
-    Fortran code:
-    w0 = 0
-    rho(:,:) = -999.0_r8
-    do kk = top_lev, pver
-       do icol = 1, ncol
-          rho(icol,kk) = pmid(icol,kk)/(rair*temperature(icol,kk))
-          w0(icol,kk) = -1._r8*omega(icol,kk)/(rho(icol,kk)*gravit)
-       enddo
-    enddo */
+    });
+  });
 
-    //---------------------------------------------------------
-    //Compute TKE using "w_sec"
-    //---------------------------------------------------------
-
-    /*
-    NOTE: All the inputs are available to compute tke
-    Fortran code:
-    tke(:ncol,:) = (3._r8/2._r8)*w_sec(:ncol,:)
-    */
+  //---------------------------------------------------------
+  //Compute TKE using "w_sec"
+  //---------------------------------------------------------
+  Kokkos::parallel_for(team_policy, KOKKOS_LAMBDA(const haero::ThreadTeam &team) {
+    const int icol = team.league_rank();
+    // FIXME Is this the correct boundary condition for tke at the surface?
+    tke_(icol,nlev_) = 0;
+    Kokkos::parallel_for(Kokkos::TeamThreadRange(team, 0u, nlev_), KOKKOS_LAMBDA(int kk) { 
+      tke_(icol,kk) = (3.0/2.0)*w_sec_(icol,kk);
+    });
+  });
    //---------------------------------------------------------
    // Compute subgrid scale velocities(wsub, wsig and wsubice)
    //---------------------------------------------------------
   
-    // More refined computation of sub-grid vertical velocity
-    // Set to be zero at the surface by initialization.
-    /*
-    NOTE: All the inputs are available to compute wsub, wsig and wsubice
-          "min_max_bound" is present in MAM4xx  utils
-
-    Fortran code:
-    wsub(:ncol,:top_lev-1)  = wsubmin
-    wsubice(:ncol,:top_lev-1) = 0.001_r8
-    wsig(:ncol,:top_lev-1)  = 0.001_r8
-
-    do kk = top_lev, pver
-       do icol = 1, ncol
-          wsub(icol,kk)  = sqrt(0.5_r8*(tke(icol,kk) + tke(icol,kk+1))*(2._r8/3._r8))
-          wsig(icol,kk)  = min_max_bound(0.001_r8, 10._r8, wsub(icol,kk))
-          wsubice(icol,kk) = min_max_bound(0.2_r8, 10._r8, wsub(icol,kk))
-          wsub(icol,kk)  = max(wsubmin, wsub(icol,kk))
-       end do
-    end do
-    */
-
+  // More refined computation of sub-grid vertical velocity
+  // Set to be zero at the surface by initialization.
+  Kokkos::parallel_for(team_policy, KOKKOS_LAMBDA(const haero::ThreadTeam &team) {
+    const int icol = team.league_rank();
+    Kokkos::parallel_for(Kokkos::TeamThreadRange(team, 0u, top_lev_), KOKKOS_LAMBDA(int kk) { 
+      wsub_(icol,kk)  = wsubmin;
+      wsubice_(icol,kk) = 0.001;
+      wsig_(icol,kk)  = 0.001;
+    });
+    Kokkos::parallel_for(Kokkos::TeamThreadRange(team, top_lev_, nlev_), KOKKOS_LAMBDA(int kk) { 
+      wsub_(icol,kk)  = haero::sqrt(0.5*(tke_(icol,kk) + tke_(icol,kk+1))*(2.0/3.0));
```

Comment:
@singhbalwinder , code on line 248 averages tke on level kk and level kk+1 in order to get a value of wsub at the cell center which would make tke a cell boundary value and not a cell center value.  But in the computation of tke at line 231 it would be at cell centers.  Is tke cell center or cell boundary value?

Reply:
TKE seems cell-centered since it is derived from w_sec, which should be cell-centered. I suspect `wsub` may be at boundaries. I will confirm this and get back to you.

Reply:
My original assumption was wrong, `w_sec` is at the boundary level. That means TKE is at the interfaces and wsub is cell-centered. Thanks for your question.